lib/sdk/content/page-mod.js

```diff
@@ -0,0 +1,249 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+"use strict";
+
+module.metadata = {
+  "stability": "stable"
+};
+
+const { getAttachEventType } = require('../content/utils');
+const { Class } = require('../core/heritage');
+const { Disposable } = require('../core/disposable');
+const { WeakReference } = require('../core/reference');
+const { WorkerChild } = require('./worker-child');
+const { EventTarget } = require('../event/target');
+const { on, emit, once, setListeners } = require('../event/core');
+const { on: domOn, removeListener: domOff } = require('../dom/events');
+const { isRegExp, isUndefined } = require('../lang/type');
+const { merge } = require('../util/object');
+const { isBrowser, getFrames } = require('../window/utils');
+const { getTabs, getTabContentWindow, getTabForContentWindow,
+        getURI: getTabURI } = require('../tabs/utils');
+const { ignoreWindow } = require('../private-browsing/utils');
+const { Style } = require("../stylesheet/style");
+const { attach, detach } = require("../content/mod");
+const { has, hasAny } = require("../util/array");
+const { Rules } = require("../util/rules");
+const { List, addListItem, removeListItem } = require('../util/list');
+const { when } = require("../system/unload");
+const { uuid } = require('../util/uuid');
+const { frames, process } = require('../remote/child');
+const { Cc, Ci, Cu } = require('chrome');
+
+const pagemods = new Map();
+const styles = new WeakMap();
+let styleFor = (mod) => styles.get(mod);
+
+// Bind observer
+const { addObserver, removeObserver } = Cc['@mozilla.org/observer-service;1'].
+                                          getService(Ci.nsIObserverService);
```

Comment:
Right, this is probably worthy of a comment and/or follow-up bug, but, until SDK add-ons are marked multiprocessCompatible registering for the document-element-inserted observer notification causes the add-on shims to forward those events from the child process. When it does so the subject parameter is a CPOW to the object in the child and any attempt to use it results in blocking cross-process messaging, part of what we're trying to avoid here. `sdk/system/events` triggers this when it does a bunch of checks on the subject to try to convert it back to an SDK object (https://github.com/mozilla/addon-sdk/blob/master/lib/sdk/system/events.js#L66). So I bypassed that by just using the raw observer service and then down in onContentWindow the first check is whether subject is a CPOW which costs nothing and we can bail out then.

It would be possible to use `sdk/core/observer` without adding extra CPOW traffic but I think that involves declaring a new class and creating an instance of it which seemed overkill.

We might be able to avoid the observer service entirely and use DOM events instead but I was having difficulty getting one or two tests passing in that case and I wanted to get this in before trying harder with that.